core/services/ocrcommon/data_source.go

```diff
@@ -176,19 +199,90 @@ func (ds *inMemoryDataSource) parse(finalResult pipeline.FinalResult) (*big.Int,
 
 // Observe without saving to DB
 func (ds *inMemoryDataSource) Observe(ctx context.Context, timestamp ocr2types.ReportTimestamp) (*big.Int, error) {
-	_, finalResult, err := ds.executeRun(ctx, ObservationTimestamp{
+	_, trrs, err := ds.executeRun(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	finalResult := trrs.FinalResult(ds.lggr)
+	setEATelemetry(ds, finalResult, trrs, ObservationTimestamp{
 		Round:        timestamp.Round,
 		Epoch:        timestamp.Epoch,
 		ConfigDigest: timestamp.ConfigDigest.Hex(),
 	})
-	if err != nil {
-		return nil, err
-	}
+
 	return ds.parse(finalResult)
 }
 
+// inMemoryDataSourceCache is a time based cache wrapper for inMemoryDataSource.
+// If cache update is overdue Observe defaults to standard inMemoryDataSource behaviour.
+type inMemoryDataSourceCache struct {
+	*inMemoryDataSource
+	cacheExpiration time.Duration
+	mu              *sync.RWMutex
```

Comment:
this doesn't need to be a pointer. a zero value works without initialization

`mu sync.RWMutex` and then you can remove the explicit initialization in other call sites. i.e. omit setting mu when calling &inMemoryDataSourceCache{..}

`The zero value for a RWMutex is an unlocked mutex.` https://pkg.go.dev/sync#RWMutex

Reply:
You can always make a pointer to the value, and the compiler even does that for you automatically when you call methods with pointer receivers.